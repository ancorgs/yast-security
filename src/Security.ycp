/**
 * File:	modules/Security.ycp
 * Package:	Security configuration
 * Summary:	Data for the security configuration
 * Authors:	Michal Svec <msvec@suse.cz>
 *
 * $Id$
 */

{

module "Security";
textdomain "security";

import "Progress";
include "security/levels.ycp";
include "security/routines.ycp";

/**
 * All security settings
 */
global map Settings = $[
    "CONSOLE_SHUTDOWN"			: "reboot",
    "CWD_IN_ROOT_PATH"			: "yes",
    "CWD_IN_USER_PATH"			: "yes",
    "DISPLAYMANAGER_REMOTE_ACCESS"	: "no",
    "FAILLOG_ENAB"			: "yes",
    "FAIL_DELAY"			: "3",
    "GID_MAX"				: "60000",
    "GID_MIN"				: "101",
    "KDM_SHUTDOWN"			: "all",
    "LASTLOG_ENAB"			: "yes",
    "OBSCURE_CHECKS_ENAB"		: "yes",
    "PASSWD_ENCRYPTION"			: "des",
    "PASSWD_USE_CRACKLIB"		: "yes",
    "PASS_MAX_DAYS"			: "99999",
    "PASS_MAX_LEN"			: "7",
    "PASS_MIN_DAYS"			: "0",
    "PASS_MIN_LEN"			: "5",
    "PASS_WARN_AGE"			: "7",
    "PERMISSION_SECURITY"		: "secure",
    "RUN_UPDATEDB_AS"			: "nobody",
    "UID_MAX"				: "60000",
    "UID_MIN"				: "100",
];

/**
 * Security settings locations
 */
map Locations = $[
    ".etc.login_defs" : [
	"FAILLOG_ENAB",
	"FAIL_DELAY",
	"GID_MAX",
	"GID_MIN",
	"LASTLOG_ENAB",
	"OBSCURE_CHECKS_ENAB",
	"PASS_MAX_DAYS",
	"PASS_MAX_LEN",
	"PASS_MIN_DAYS",
	"PASS_MIN_LEN",
	"PASS_WARN_AGE",
	"UID_MAX",
	"UID_MIN",
    ],
    ".sysconfig.suseconfig" : [
	"CWD_IN_ROOT_PATH",
	"CWD_IN_USER_PATH",
    ],
    ".sysconfig.displaymanager" : [
	"DISPLAYMANAGER_REMOTE_ACCESS",
	"KDM_SHUTDOWN",
    ],
    ".sysconfig.security" : [
	"PERMISSION_SECURITY",
    ],
    ".sysconfig.locate" : [
	"RUN_UPDATEDB_AS",
    ],
];

/**
 * Remaining settings:
 * - CONSOLE_SHUTDOWN (/etc/inittab)
 * - PASSWD_ENCRYPTION (/etc/pam?)
 * - PASSWD_USE_CRACKLIB (/etc/pam?)
 */

/**
 * Abort function
 * return boolean return true if abort
 */
global block AbortFunction = nil;

/**
 * Check for pending Abort press
 * @return true if pending abort
 */
global define boolean PollAbort() ``{
    return UI::PollInput() == `abort;
}

/**
 * Abort function
 * @return blah blah lahjk
 */
global define boolean Abort() ``{
    if(AbortFunction != nil)
	return eval(AbortFunction) == true;
    return false;
}

/**
 * Data was modified?
 */
global boolean modified = false;

global boolean proposal_valid = false;
global boolean write_only = false;

/**
 * Data was modified?
 * @return true if modified
 */
global define boolean Modified() ``{
    y2debug("modified=%1",modified);
    return modified;
};

/**
 * Read all security settings
 * @return true on success
 */
global define boolean Read() ``{

    /* Security read dialog caption */
    string caption = _("Initializing security configuration");
    integer steps = 3;

    integer sl = 0*1000;
    sleep(sl);

    Progress::New( caption, " ", steps, [
	    /* Progress step 1/3 */
	    _("Read security settings"),
	    /* Progress step 2/3 */
	    _("Read inittab settings"),
	    /* Progress step 3/3 */
	    _("Read PAM settings"),
	], [
	    /* Progress stage 1/4 */
	    _("Reading security settings"),
	    /* Progress stage 2/4 */
	    _("Reading inittab settings"),
	    /* Progress stage 3/4 */
	    _("Reading PAM settings"),
	    /* Progress stage 4/4 */
	    _("Finished")
	],
	""
    );

    Settings = $[];

    /* Read security settings */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    mapmap(string file, list vars, Locations, ``{
	maplist(string var, vars, ``{
	    string val = "";

	    string filename = nil;
	    if(issubstring(file, "sysconfig")) {
		filename = "/etc" + mergestring(splitstring(file, "."), "/");
		y2debug("filename=%1", filename);
	    }

	    if(filename == nil || SCR::Read(.target.size, filename) > 0) {
		val = SCR::Read(topath(file + "." + var));
		y2debug("Reading: %1 (%2)", topath(file + "." + var), val);
	    }
	    if(val == nil) val = "";
	    Settings[var] = val;
	});
	return [0,0];
    });
    y2debug("Settings=%1", Settings);

    /* Read inittab settings */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    list inittab = SCR::Dir(.etc.inittab);
    if(contains(inittab, "ca")) {
	string ca = SCR::Read(.etc.inittab.ca);
	if(issubstring(ca, "/bin/true") || issubstring(ca, "/bin/false"))
	    change(Settings, "CONSOLE_SHUTDOWN", "ignore");
	else if(issubstring(ca, "reboot") || issubstring(ca, " -r"))
	    change(Settings, "CONSOLE_SHUTDOWN", "reboot");
	else if(issubstring(ca, "halt") || issubstring(ca, " -h"))
	    change(Settings, "CONSOLE_SHUTDOWN", "halt");
	else {
	    y2error("Unknown ca status: %1", ca);
	    change(Settings, "CONSOLE_SHUTDOWN", "ignore");
	}
    }
    else
	change(Settings, "CONSOLE_SHUTDOWN", "ignore");

    y2debug("Settings=%1", Settings);

    /* Read pam settings */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    /* pam stuff */
    map pam_unix_map = $[];
    list pam_list = SCR::Read(.pam.all.password.pam_unix2);
    if(pam_list != [] && pam_list != nil)
	pam_unix_map = select(pam_list,0,$[]);

    map pam_check_map = $[];
    pam_list = SCR::Read(.pam.all.password.pam_pwcheck);
    if(pam_list != [] && pam_list != nil)
	pam_check_map = select(pam_list,0,$[]);

    /* md5 & blowfish support */
    boolean md5_support = contains(splitstring(lookup(pam_unix_map , "arguments", ""), " "), "md5");
    boolean blowfish_support = contains(splitstring(lookup(pam_unix_map , "arguments", ""), " "), "blowfish");
    string method = "";
    if (md5_support)
	method = "md5";
    else if (blowfish_support)
	method = "blowfish";
    else
	method = "des";
    change(Settings,"PASSWD_ENCRYPTION", method);

    /* use cracklib? */
    boolean pam_support = contains(splitstring(lookup(pam_check_map , "arguments", ""), " "), "use_cracklib");
    change(Settings,"PASSWD_USE_CRACKLIB", pam_support?"yes":"no");
    y2debug("Settings=%1", Settings);

    /* Local permissions hack */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    string perm = Settings["PERMISSION_SECURITY"]:"";
    if(issubstring(perm, "easy")) perm = "easy";
    else if(issubstring(perm, "paranoid")) perm = "paranoid";
    else if(issubstring(perm, "secure")) perm = "secure";
    else perm = "secure";
    change(Settings, "PERMISSION_SECURITY", perm);
    y2debug("Settings=%1", Settings);

    sleep(sl);
    if(Abort()) return false;
    modified = false;
    return true;
}

/**
 * Write all security settings
 * @return true on success
 */
global define boolean Write() ``{

    if(!modified) return true;
    y2milestone("Writing configuration");

    /* Security read dialog caption */
    string caption = _("Saving security configuration");
    integer steps = 4;

    integer sl = 0*1000;
    sleep(sl);

    Progress::New( caption, " ", steps, [
	    /* Progress step 1/4 */
	    _("Write security settings"),
	    /* Progress step 2/4 */
	    _("Write inittab settings"),
	    /* Progress step 3/4 */
	    _("Write PAM settings"),
	    /* Progress step 4/4 */
	    _("Run SuSEconfig"),
	], [
	    /* Progress stage 1/5 */
	    _("Writing security settings"),
	    /* Progress stage 2/5 */
	    _("Writing inittab settings"),
	    /* Progress stage 3/5 */
	    _("Writing PAM settings"),
	    /* Progress stage 4/5 */
	    _("Running SuSEconfig"),
	    /* Progress stage 5/5 */
	    _("Finished")
	],
	""
    );

    /* Write security settings */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    y2debug("Settings=%1", Settings);
    Settings["PERMISSION_SECURITY"] = Settings["PERMISSION_SECURITY"]:"" + " local";

    mapmap(string file, list vars, Locations, ``{
	maplist(string var, vars, ``{
	    string val = Settings[var]:"";
	    if(val == nil) val = "";
	    SCR::Write(topath(file + "." + var), val);
	});
	return [0,0];
    });

    maplist(string file, mapkeys(Locations), ``{
	SCR::Write(topath(file), nil);
    });

    /* Write inittab settings */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    string ca = Settings["CONSOLE_SHUTDOWN"]:"ignore";
    if(ca == "reboot")
	SCR::Write(.etc.inittab.ca, ":ctrlaltdel:/sbin/shutdown -r -t 4 now");
    else if(ca == "halt")
	SCR::Write(.etc.inittab.ca, ":ctrlaltdel:/sbin/shutdown -h -t 4 now");
    else
	SCR::Write(.etc.inittab.ca, ":ctrlaltdel:/bin/true");
    SCR::Write(.etc.inittab, nil);

    /* Write pam settings */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    /* pam stuff */
    string encr = Settings["PASSWD_ENCRYPTION"]:"des";

    if(encr != "md5") {
	SCR::Write(.pam.all.password.pam_unix2, "-md5");
	SCR::Write(.pam.all.password.pam_pwcheck, "-md5");
	/* Remove old config options */
	SCR::Write(.pam.login.password.pam_pwcheck, "-md5");
	SCR::Write(.pam.login.password.pam_unix, "-md5");
	SCR::Write(.pam.passwd.password.pam_pwcheck, "-md5");
	SCR::Write(.pam.passwd.password.pam_unix, "-md5");
	SCR::Write(.pam.sshd.password.pam_pwcheck, "-md5");
	SCR::Write(.pam.sshd.password.pam_unix, "-md5");
	SCR::Write(.pam.rlogin.password.pam_pwcheck, "-md5");
	SCR::Write(.pam.rlogin.password.pam_unix, "-md5");
    }
    if(encr != "blowfish") {
	SCR::Write(.pam.all.password.pam_unix2, "-blowfish");
	SCR::Write(.pam.all.password.pam_pwcheck, "-blowfish");
	/* Remove old config options */
	SCR::Write(.pam.login.password.pam_pwcheck, "-blowfish");
	SCR::Write(.pam.login.password.pam_unix, "-blowfish");
	SCR::Write(.pam.passwd.password.pam_pwcheck, "-blowfish");
	SCR::Write(.pam.passwd.password.pam_unix, "-blowfish");
	SCR::Write(.pam.sshd.password.pam_pwcheck, "-blowfish");
	SCR::Write(.pam.sshd.password.pam_unix, "-blowfish");
	SCR::Write(.pam.rlogin.password.pam_pwcheck, "-blowfish");
	SCR::Write(.pam.rlogin.password.pam_unix, "-blowfish");
    }

    /* md5 support */
    if(encr == "md5") {
       SCR::Write(.pam.all.password.pam_unix2, "+md5");
       SCR::Write(.pam.all.password.pam_pwcheck, "+md5");
    }
    /* blowfish support */
    else if(encr == "blowfish") {
       SCR::Write(.pam.all.password.pam_unix2, "+blowfish");
       SCR::Write(.pam.all.password.pam_pwcheck, "+blowfish");
    }

    /* use cracklib? */
    string pam_string = (Settings["PASSWD_USE_CRACKLIB"]:"no" == "yes") ? "+use_cracklib" : "-use_cracklib";
    SCR::Write(.pam.all.password.pam_pwcheck, pam_string);

    /* Finish him: SuSEconfig */
    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    SCR::Execute(.target.bash, "/sbin/SuSEconfig --module xdm");
    SCR::Execute(.target.bash, "/sbin/SuSEconfig --module kde3");
    SCR::Execute(.target.bash, "/sbin/SuSEconfig --module permissions");
    SCR::Execute(.target.bash, "/sbin/SuSEconfig --module profiles");

    sleep(sl);
    if(Abort()) return false;
    Progress::NextStage();

    sleep(sl);
    if(Abort()) return false;
    modified = false;
    return true;
}

/**
 * Get all security settings from the first parameter
 * (For use by autoinstallation.)
 * @param settings The YCP structure to be imported.
 * @return boolean True on success
 */
global define boolean Import(map settings) ``{
    Settings = eval(settings);
}

/**
 * Dump the security settings to a single map
 * (For use by autoinstallation.)
 * @return map Dumped settings (later acceptable by Import ())
 */
global define map Export() ``{
    return eval(Settings);
}

/**
 * Create a textual summary and a list of unconfigured cards
 * @return summary of the current configuration
 */
global define list Summary() ``{

    /* Determine current settings */
    any current = `custom;
    maplist(string key, map level, Levels, ``{
	y2debug("%1=%2", key, level);
	if(level == Settings)
	    current = key;
    });
    y2debug("%1=%2", "curren", Settings);

    /* Summary text */
    string summary = _("Current security level: Custom settings");
    if(current != `custom)
	/* Summary text */
	summary = sformat(_("Current security level: %1"), LevelsNames[current]:"");

    return [ summary, [] ];
}

/**
 * Create an overview table with all configured cards
 * @return table items
 */
global define list Overview() ``{
    return [];
}

/* EOF */
}
