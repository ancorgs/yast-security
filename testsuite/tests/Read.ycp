/*
 * YaST2: Modules testsuite
 *
 * Description:
 *   Testsuite for the security module
 *
 * Authors:
 *   Michal Svec <msvec@suse.cz>
 *
 * $Id$
 *
 * testedfiles: Security.ycp PamSettings.ycp Pam.ycp
 */

{

include "testsuite.ycp";
import "Security";

map scr_info = $[
    "sysconfig" : $[
	"security" : $[
	    "PERMISSION_SECURITY"	:	"paranoid",
	    "DISABLE_RESTART_ON_UPDATE" :	"r13",
	    "DISABLE_STOP_ON_REMOVAL"	:	"r14",
	],
	"suseconfig" : $[
	    "CWD_IN_ROOT_PATH"		:	"r2",
	    "CWD_IN_USER_PATH"		:	"r3",
	],
	"displaymanager" : $[
	    "DISPLAYMANAGER_SHUTDOWN"		:	"r3",
	    "DISPLAYMANAGER_REMOTE_ACCESS" :	"r9",
	    "DISPLAYMANAGER_ROOT_LOGIN_REMOTE"	: "r16",
	    "DISPLAYMANAGER_XSERVER_TCP_PORT_6000_OPEN"	: "r17",
	],
	"locate" : $[
	    "RUN_UPDATEDB_AS"		:	"r7",
	],
	"sysctl" : $[
	    "ENABLE_SYSRQ"		:	"r8",
	    "IP_TCP_SYNCOOKIES"		:	"r9",
	    "IP_FORWARD"		:	"r10",
	    "IPV6_FORWARD"		:	"r11",
	],
	"clock" : $[
	    "SYSTOHC"			:	"r12",
	],
	"cron" : $[
	    "SYSLOG_ON_NO_ERROR"	:	"r15",
	],
	"mail" : $[
	    "SMTPD_LISTEN_REMOTE"	:	"r18",
	],
    ],
    "etc" : $[
	"login_defs" : $[
	    "FAIL_DELAY"		:	"l2",
	    "GID_MAX"			:	"l3",
	    "GID_MIN"			:	"l4",
	    "LASTLOG_ENAB"		:	"l5",
	    "PASS_MAX_DAYS"		:	"l7",
	    "PASS_MIN_DAYS"		:	"l9",
	    "PASS_MIN_LEN"		:	"l10",
	    "PASS_WARN_AGE"		:	"l11",
	    "UID_MAX"			:	"l12",
	    "UID_MIN"			:	"l13",
	    "SYSTEM_UID_MAX"		:	"l14",
	    "SYSTEM_UID_MIN"		:	"l15",
	    "SYSTEM_GID_MAX"		:	"l16",
	    "SYSTEM_GID_MIN"		:	"l17",
	    "USERADD_CMD"		:	"l18",
	    "USERDEL_PRECMD"		:	"l19",
	    "USERDEL_POSTCMD"		:	"l20",
	],
	"inittab" : $[
	    "ca"			:	":ctrlaltdel:/sbin/shutdown -r -t 4 now"
	],
	"default" : $[
	    "passwd" : $[
		"crypt"			:	"md5",
		"CRYPT_FILES"		:	"blowfish",
	    ],
	],
    ],
    "target" : $[ "size" : 1 ],
    "pam" : $[ "passwd" : $[ "password" : $[
	"pam_unix" : [
	    $[ "arguments" : "nullok use_first_pass use_authtok" ]
	]
    ]]]
];

map E	= $[
    "target" : $[
	"bash_output" : $[]
    ]
];

TEST(``(Security::Read()),[scr_info,$[],E],nil);

}
