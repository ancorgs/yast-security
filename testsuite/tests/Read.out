Read	.etc.login_defs.FAILLOG_ENAB "l1"
Read	.etc.login_defs.FAIL_DELAY "l2"
Read	.etc.login_defs.GID_MAX "l3"
Read	.etc.login_defs.GID_MIN "l4"
Read	.etc.login_defs.LASTLOG_ENAB "l5"
Read	.etc.login_defs.PASS_MAX_DAYS "l7"
Read	.etc.login_defs.PASS_MIN_DAYS "l9"
Read	.etc.login_defs.PASS_WARN_AGE "l11"
Read	.etc.login_defs.UID_MAX "l12"
Read	.etc.login_defs.UID_MIN "l13"
Read	.etc.login_defs.SYSTEM_UID_MAX "l14"
Read	.etc.login_defs.SYSTEM_UID_MIN "l15"
Read	.etc.login_defs.SYSTEM_GID_MAX "l16"
Read	.etc.login_defs.SYSTEM_GID_MIN "l17"
Read	.etc.login_defs.USERADD_CMD "l18"
Read	.etc.login_defs.USERDEL_PRECMD "l19"
Read	.etc.login_defs.USERDEL_POSTCMD "l20"
Read	.target.size "/etc/sysconfig/displaymanager" 1
Read	.sysconfig.displaymanager.DISPLAYMANAGER_REMOTE_ACCESS "r9"
Read	.target.size "/etc/sysconfig/displaymanager" 1
Read	.sysconfig.displaymanager.KDM_SHUTDOWN "r3"
Read	.target.size "/etc/sysconfig/locate" 1
Read	.sysconfig.locate.RUN_UPDATEDB_AS "r7"
Read	.target.size "/etc/sysconfig/security" 1
Read	.sysconfig.security.PERMISSION_SECURITY "paranoid"
Read	.target.size "/etc/sysconfig/suseconfig" 1
Read	.sysconfig.suseconfig.CWD_IN_ROOT_PATH "r2"
Read	.target.size "/etc/sysconfig/suseconfig" 1
Read	.sysconfig.suseconfig.CWD_IN_USER_PATH "r3"
Read	.target.size "/etc/sysconfig/sysctl" 1
Read	.sysconfig.sysctl.ENABLE_SYSRQ "r8"
Dir	.etc.inittab: ["ca"]
Read	.etc.inittab.ca ":ctrlaltdel:/sbin/shutdown -r -t 4 now"
Dir	.etc.security.section: []
Read	.etc.security.v."pam_unix2"."password" nil
Dir	.etc.security.section: []
Read	.etc.security.v."pam_pwcheck"."password" nil
Return	true
